SIMSBIOHUB-648: Create Sampling Period Page

mauberti-bc opened the pull request:
## Links to Jira Tickets

- {Include a link to all applicable Jira tickets}

## Description of Changes

The main goal of this PR is to separate the sampling site and sampling period CRUD operations. Previously sampling periods were created/updated as part of creating/updating a sampling site. Now, sampling periods have their own create/edit pages/components.

All of the changes are around that concept, handling all of the code that naturally broke as a result of the related database schema changes, and workflow changes, etc.

## Testing Notes

### New DB Schema
- 1 new migration
  - See JSDoc notes in the migration.
- 1 new procedure
  -  Check constraint for the sample_period table.

Picture of the relevant tables, and how they look after running the migration:
<img src="https://github.com/user-attachments/assets/4fd8c157-d613-4b17-9ab5-687df555bb56" width="700">


### New Functionality
- Separate table/page for creating, editing, deleting sampling periods (via the Manage Sampling Information page).
  - New endpoints/services/repositories/interfaces

### Updated Functionality
- Updated the 3 sampling columns in the Observations table. Behaviour is similar to before, but underlying data is different.
- Updated the sampling side bar (from the Manage Observations page). Behaviour is similar to before, but underlying data is different.
- Updated Analytics tab (below the map on the Survey page). Behaviour is similar to before, but underlying data is different.
- Updated the sampling information portion of the existing Observation CSV import backend code. The rest of the Observation CSV import code should work as before.

mauberti-bc committed:
sampling period table on manage page

mauberti-bc committed:
remove periods tab from sampling technique table

mauberti-bc committed:
add periods

mauberti-bc committed:
sampling period create page

mauberti-bc committed:
remove methods from creating sample site workflow

mauberti-bc committed:
api changes

mauberti-bc committed:
fix test

mauberti-bc committed:
Merge branch 'dev' of github.com:bcgov/biohubbc into SIMSBIOHUB-648

mauberti-bc committed:
edit sampling period

mauberti-bc committed:
Merge branch 'dev' into SIMSBIOHUB-648

mauberti-bc committed:
move sample-method/{methodId}/sample-period api to survey/sample-period

mauberti-bc committed:
fix update endpoint

mauberti-bc committed:
add missing backend tests

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into SIMSBIOHUB-648

NickPhura commented:
A few things I noticed:

---

When creating a period, if you select a site, and then remove the period from it, the validation is probably failing because the form won't submit, but there is no error message shown.

![image](https://github.com/user-attachments/assets/fb57aba6-37db-4907-8f1d-776677af9d7e)

---

When editing an existing sample period, I can't select or change the sample sites.

![image](https://github.com/user-attachments/assets/f8f46124-2d85-43ab-8ada-3bb29a335be7)

---

Nit: I can save a period with a start and end date that are exactly the same (with no time). 

![image](https://github.com/user-attachments/assets/bc89ce88-3a42-4e75-bea6-e5006536ffa0)

If I pick the same times, then the save stops working, though there is no error message about why.

---

I have 1 sampling site, 1 technique, and 3 periods. The techniques aren't being combined in the observations table when adding a record:

![image](https://github.com/user-attachments/assets/782763e7-68a9-4cf9-baf8-603cad9e7e4d)

mauberti-bc commented:
- I disallowed changing the site because at that point you may as well create a new period for the correct site and delete the wrong one. Seems along the lines of turning a skateboard into a car instead of just buying a car. Maybe worth allowing though? 
- We do want to allow a period that starts and stops on the same day with no time, in case the time is not known. I agree this seems off but is realistic data. (The duration column is null for those periods because there isn't an obvious way to represent the unknown duration)

NickPhura committed:
Create/Edit Sampling Period forms working.

NickPhura committed:
WIP: save point

NickPhura committed:
Manage observation Sampling controls working

NickPhura committed:
Refactor sampling information utils

NickPhura committed:
Fix analytics backend.
Fix some api tests.
Remove some deprecated code.

NickPhura committed:
Remove some deprecated code.

NickPhura committed:
Update most sampling "location" to sampling "site"

mauberti-bc commented:
![image](https://github.com/user-attachments/assets/5d9f8066-2062-4e6c-b3e3-c4f0e85fb7a1)
There's a failing request when expanding a sampling site in the left side panel of the observations page. Otherwise haven't found any issues

NickPhura committed:
Update survey_sample_period table to allow all fields to be null.
Update api and app to account for null survey_sample_period fields.

mauberti-bc commented:
* No observations are being returned on the observations table page, but the request succeeds
* When you select a site or technique, then remove that site or technique from the form, that option does not return to the autocomplete options (you have to refresh the autocomplete for it to show up). 
* We should also load the sampling site and technique autocompletes when the component mounts, before any input.

mauberti-bc committed:
change form text & add help markdown

NickPhura committed:
Fixed issue preventing the survey list under a project from not loading always.
Fixed issue preventing observations from loading.
Fixed issue in seed.

NickPhura committed:
Merge branch 'SIMSBIOHUB-648' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-648

NickPhura committed:
Fix existing unit tests

NickPhura committed:
Update find periods to allow arrays of sample site and technique ids.
Update create/edit periods to enforce site, technique, and period information.
UI tweaks to create/edit periods pages.
Enforce site/technique/period validation on observations table.
Update sample sites left list, to show technique/period details.

NickPhura committed:
Fix unit tests

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into SIMSBIOHUB-648

NickPhura committed:
Fix analytics API.
Tweak analytics frontend (disable sorting which is not yet supported)

NickPhura committed:
Update analytics interface

NickPhura committed:
Add database trigger to check survey sample period minimum data requirements

NickPhura committed:
Add endpoint unit tests.
Fix some existing req.query parsing logic.

NickPhura committed:
Add prop to automatically run the search when using the generic AutocompleteSearchField.tsx.

NickPhura committed:
Add endpoint unit tests.
Remove unused api param from delete observations endpoint.
Minor tweak to find endpoint filter type (allow them to be undefined as they were meant to be)

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into SIMSBIOHUB-648

NickPhura committed:
- Add new migration to drop n8n schema.
- Rename/refactor sample method migrations to come after the upgrade postgres migration.

NickPhura committed:
ignore-skip

NickPhura committed:
Fix bug in findTechniques repo function

NickPhura committed:
Enhancements

MacDeluca commented:
Heres some screenshots of some of the issues I found:

1. When deleting the sampling site from the "Manage Observations" page it throws a failed to execute SQL error: column "survey_sample_site_id" does not exist
![image](https://github.com/user-attachments/assets/091d7224-e151-4a36-94bc-c10bc13687a7)

2. The sampling techniques table renders the wrong page when the rows overflow into the next page. (renders page 2 vs rendering page 1) when 11 sampling techniques http://localhost:7100/admin/projects/2/surveys/4/sampling
![image](https://github.com/user-attachments/assets/89ff2ddb-b1b1-494f-8943-c978b2e9a500)


3. The survey page Sampling Information (Technique tab) says 11 records but renders nothing on page 2 http://localhost:7100/admin/projects/2/surveys/4/details
![image](https://github.com/user-attachments/assets/cc2a9e72-53d2-4ca0-ae56-ce90b1bc605f)

NickPhura committed:
Address PR feedback
- Fix techniques pagination (on both survey technique page, and manage sampling information page).
- Fix delete sample sites sql error.
- Update APIError -> HTTPError code.

NickPhura commented:
> Heres some screenshots of some of the issues I found:

Pushed up some changes that should address those 3 issues.

MacDeluca commented on database/src/procedures/tr_before_survey_sample_period_check_minimum_data_requirements.ts:
```diff
@@ -0,0 +1,45 @@
+  await knex.raw(`--sql
+    SET search_path = 'biohub';
+
+    CREATE OR REPLACE FUNCTION biohub.check_survey_sample_period_minimum_data_requirements()
```
suggestion: We might be able to replace this trigger with a CHECK contraint using `num_nonnulls`
https://stackoverflow.com/questions/15178859/postgres-constraint-ensuring-one-column-of-many-is-present

NickPhura replied:
Will try it out

NickPhura replied:
pushed up new migration (20250117000001_survey_sample_period_check_constraints) and dropped the trigger

MacDeluca commented on api/src/repositories/sample-site-repository/utils.ts:
```diff
@@ -0,0 +1,190 @@
+        'sss.name',
+        'sss.description',
+        'sss.geojson',
+        // 'technique',
```
💀

NickPhura replied:
Removed

MacDeluca commented on api/src/services/geo-service.ts:
```diff
@@ -228,6 +230,13 @@ export class GeoService {
+    axios.interceptors.response.use(
+      (response) => response,
+      (error) => {
+        throw new ApiGeneralError('Failed to geospatial data from external service.', [formatAxiosError(error)]);
```
grammar mistake: `Failed to retrieve geospatial data from external service`

NickPhura replied:
Fixed!

NickPhura committed:
Move survey_sample_period constraint into check constraint on the table (instead of separate trigger).

NickPhura committed:
Update constraint

MacDeluca commented on app/src/features/surveys/sampling-information/techniques/edit/EditTechniquePage.tsx:
```diff
@@ -71,6 +72,7 @@ export const EditTechniquePage = () => {
     description: technique?.description ?? null,
     distance_threshold: technique?.distance_threshold ?? null,
     method_lookup_id: technique?.method_lookup_id ?? null,
+    method_response_metric_id: 1, // NICK TODO: This is a placeholder value.
```
nit: Still needed? Can set to `-1` to prevent issues later.

NickPhura replied:
Will check with @mauberti-bc. The technique table has a non-null foreign key to the method_response_metric table, but it appears the UI no longer has the field for entering that value. Setting to `1` sets it to `count` (albeit unsafely since the ids aren't stable). Probably need to either re-add that form control so it can be set by the user, or else make that field nullable and pass null here instead. Or get the value from the codes, and still set it to the id of the `count` code, which will be the same result but safer.

Note to self:
- Check both edit and create technique pages
- Check `ObservationCountDataGridEditCell` as it is also setting this value to `1`

NickPhura replied:
Fixed. Added proper form control to populate this field.

MacDeluca commented on app/src/features/surveys/sampling-information/techniques/create/CreateTechniquePage.tsx:
```diff
@@ -63,6 +65,7 @@ export const CreateTechniquePage = () => {
       const createTechniqueRequestData: ICreateTechniqueRequest = {
         ...values,
         distance_threshold: values.distance_threshold || null,
+        method_response_metric_id: 1, // NICK TODO: This is a placeholder value.
```
nit: Same as above comment

NickPhura replied:
Fixed. Added proper form control to populate this field.

MacDeluca commented on database/src/migrations/20250117000001_survey_sample_period_check_constraints.ts:
```diff
@@ -0,0 +1,40 @@
+      NUM_NONNULLS(survey_sample_site_id, method_technique_id, start_date) >= 1
+    );
+
+    COMMENT ON CONSTRAINT check_minimum_data_requirement ON survey_sample_period IS 'Checks that at least one of the following items is not null: survey_sample_site_id, method_technique_id, start_date + end_date. Note: This constraint only checks start_date because the check_start_and_end_date_requirement constraint already ensures that start_date and end_date are either both null or both not nul ...
```
praise: Hell yeah 🎸

NickPhura committed:
Remove unused code.
Fix error message.

NickPhura committed:
Fix sampling datagrid cell behaviour when clearing a site or technique

NickPhura committed:
Add response metric field to create/edit technique page.

NickPhura committed:
Update sample period matching logic.
Add unit tests for sample period matching functions.

NickPhura committed:
Fix date-time-utils.
Add unit tests

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into SIMSBIOHUB-648

MacDeluca commented on api/src/services/import-services/utils/datetime.ts:
```diff
@@ -21,15 +22,15 @@ interface ICaptureStub {
 export const formatTimeString = (time?: string | null): string | undefined => {
-  const fullTime = dayjs(time, 'HH:mm:ss');
+  const fullTime = dayjs(time, DefaultTimeFormat);
   const shortTime = dayjs(time, 'HH:mm');
```
nit: This can be updated to the new `DefaultTimeFormatNoSeconds`

NickPhura replied:
Updated

NickPhura committed:
Adjust sample period matching logic: Now only matches on sampling information OR observation date/time (not both).

NickPhura committed:
Update to use DefaultTimeFormatNoSeconds

NickPhura committed:
Fix placeholder spelling.

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into SIMSBIOHUB-648

MacDeluca reviewed (approved):
👍🏼
